nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-nar-utils/src/main/java/org/apache/nifi/nar/StandardExtensionDiscoveringManager.java

```diff
@@ -383,42 +388,69 @@ public InstanceClassLoader createInstanceClassLoader(final String classType, fin
+                                break;
+                            }
 
-            if (requiresInstanceClassLoading.cloneAncestorResources()) {
-                final ConfigurableComponent component = getTempComponent(classType, bundle.getBundleDetails().getCoordinate());
-                final Set<BundleCoordinate> reachableApiBundles = findReachableApiBundles(component);
+                            final NarClassLoader ancestorNarClassLoader = (NarClassLoader) ancestorClassLoader;
 
-                while (ancestorClassLoader instanceof NarClassLoader) {
-                    final Bundle ancestorNarBundle = classLoaderBundleLookup.get(ancestorClassLoader);
+                            narNativeLibDirs.add(ancestorNarClassLoader.getNARNativeLibDir());
+                            Collections.addAll(instanceUrls, ancestorNarClassLoader.getURLs());
 
-                    // stop including ancestor resources when we reach one of the APIs, or when we hit the Jetty NAR
-                    if (ancestorNarBundle == null || reachableApiBundles.contains(ancestorNarBundle.getBundleDetails().getCoordinate())
-                            || ancestorNarBundle.getBundleDetails().getCoordinate().getId().equals(NarClassLoaders.JETTY_NAR_ID)) {
-                        break;
+                            ancestorClassLoader = ancestorNarClassLoader.getParent();
+                        }
                     }
+                }
 
-                    final NarClassLoader ancestorNarClassLoader = (NarClassLoader) ancestorClassLoader;
-
-                    narNativeLibDirs.add(ancestorNarClassLoader.getNARNativeLibDir());
-                    Collections.addAll(instanceUrls, ancestorNarClassLoader.getURLs());
+                // register our new InstanceClassLoader as the shared base classloader.
+                if (baseClassLoaderKey != null && !resolvedSharedClassLoader) {
+                    // Created a shared class loader that is everything we need except for the additional URLs, as the additional URLs are instance-specific.
+                    final ClassLoader sharedClassLoader = new SharedInstanceClassLoader(instanceIdentifier, classType, instanceUrls, Collections.emptySet(), narNativeLibDirs, ancestorClassLoader);
+                    instanceClassLoader = new InstanceClassLoader(instanceIdentifier, classType, Collections.emptySet(), additionalUrls, Collections.emptySet(), sharedClassLoader);
 
-                    ancestorClassLoader = ancestorNarClassLoader.getParent();
+                    logger.debug("Creating InstanceClassLoader for type {} using newly created shared Base ClassLoader {} for component {}", type, instanceClassLoader, instanceIdentifier);
+                    sharedBaseClassloaders.putIfAbsent(baseClassLoaderKey, sharedClassLoader);
+                } else {
+                    instanceClassLoader = new InstanceClassLoader(instanceIdentifier, classType, instanceUrls, additionalUrls, narNativeLibDirs, ancestorClassLoader);
```

Comment:
I think this may need to be split into a third case... 

Right now when creating a component that has an existing isolation key, it will hit this else case and create the `InstanceClassLoader` with `instanceUrls` which has the current NAR's resources, but in this case those resources are already in the `SharedInstanceClassLoader.`

In the specific example of HDFS processors, the `SharedInstanceClassLoader` has `nifi-hadoop-nar` + `nifi-hadoop-libraries-nar`, and then the next processor that uses this same `SharedInstanceClassLoader` also gets `nifi-hadoop-nar` resources in its `InstanceClassLoader`.